base/utils/R/read.output.R

```diff
@@ -36,11 +36,11 @@ read.output <- function(runid, outdir, start.year = NA, end.year = NA, variables
   ## 'TotalResp', 'AutoResp', 'HeteroResp', 'DOC_flux', 'Fire_flux') # kgC m-2 s-1
   ## wflux = c('Evap', 'TVeg', 'Qs', 'Qsb', 'Rainf') # kgH20 m-2 d-1
   
-  # create list of *.nc years
-  nc.years <- as.vector(unlist(strsplit(list.files(path = outdir, pattern = "\\.nc$", 
+  # create list of *.nc years - look only for files formatted as YYYY.nc, the default pecan output file name standard
+  nc.years <- as.vector(unlist(strsplit(list.files(path = outdir, pattern = "^-?[[:digit:]]{4}\\.nc$", 
                                                    full.names = FALSE), ".nc")),
                         mode = "numeric") # N.B. coercing from character
-  ncfiles <- list.files(path = outdir, pattern = "\\.nc$", full.names = TRUE)
+  ncfiles <- list.files(path = outdir, pattern = "^-?[[:digit:]]{4}\\.nc$", full.names = TRUE)
```

Comment:
I would call this before `nc.years`, then pass `ncfiles` as input to `nc.years`. Otherwise, you're duplicating effort.

```r
ncfiles_sub <- list.files(outdir, ..., full.names = FALSE)
ncfiles_full <- file.path(outdir, ncfiles_sub)
nc_years <- as.vector(unlist(strsplit(ncfiles_sub)), "numeric")
# Alternatively, pull out the year using a regex capture group
nc_years <- as.numeric(gsub("^(-?[[:digit:]]{4})\\.nc", "\\1", ncfiles_sub))
```

Reply:
Note that `file.path` is vectorized, so the way it's used here will prepend `outdir` to all elements of `ncfiles_sub`.

Reply:
How about
```
outdir <- "/data/sserbin/Modeling/dvmdostem/pecan_runs/salix_run.029/out/SA-median"

ncfiles_sub <- list.files(path = outdir, pattern = "^-?[[:digit:]]{4}\\.nc$", full.names = FALSE)
> ncfiles_sub
  [1] "-1901.nc" "1901.nc"  "1902.nc"  "1903.nc"  "1904.nc"  "1905.nc"
  [7] "1906.nc"  "1907.nc"  "1908.nc"  "1909.nc"  "1910.nc"  "1911.nc"
 [13] "1912.nc"  "1913.nc"  "1914.nc"  "1915.nc"  "1916.nc"  "1917.nc"
 [19] "1918.nc"  "1919.nc"  "1920.nc"  "1921.nc"  "1922.nc"  "1923.nc"
 [25] "1924.nc"  "1925.nc"  "1926.nc"  "1927.nc"  "1928.nc"  "1929.nc"
 [31] "1930.nc"  "1931.nc"  "1932.nc"  "1933.nc"  "1934.nc"  "1935.nc"
 [37] "1936.nc"  "1937.nc"  "1938.nc"  "1939.nc"  "1940.nc"  "1941.nc"
 [43] "1942.nc"  "1943.nc"  "1944.nc"  "1945.nc"  "1946.nc"  "1947.nc"
 [49] "1948.nc"  "1949.nc"  "1950.nc"  "1951.nc"  "1952.nc"  "1953.nc"
 [55] "1954.nc"  "1955.nc"  "1956.nc"  "1957.nc"  "1958.nc"  "1959.nc"
 [61] "1960.nc"  "1961.nc"  "1962.nc"  "1963.nc"  "1964.nc"  "1965.nc"
 [67] "1966.nc"  "1967.nc"  "1968.nc"  "1969.nc"  "1970.nc"  "1971.nc"
 [73] "1972.nc"  "1973.nc"  "1974.nc"  "1975.nc"  "1976.nc"  "1977.nc"
 [79] "1978.nc"  "1979.nc"  "1980.nc"  "1981.nc"  "1982.nc"  "1983.nc"
 [85] "1984.nc"  "1985.nc"  "1986.nc"  "1987.nc"  "1988.nc"  "1989.nc"
 [91] "1990.nc"  "1991.nc"  "1992.nc"  "1993.nc"  "1994.nc"  "1995.nc"
 [97] "1996.nc"  "1997.nc"  "1998.nc"  "1999.nc"  "2000.nc"  "2001.nc"
[103] "2002.nc"  "2003.nc"  "2004.nc"  "2005.nc"  "2006.nc"  "2007.nc"
[109] "2008.nc"  "2009.nc"  "2010.nc"  "2011.nc"  "2012.nc"  "2013.nc"
[115] "2014.nc"  "2015.nc"  "2016.nc"  "2017.nc"  "2018.nc"  "2019.nc"
[121] "2020.nc"  "2021.nc"  "2022.nc"  "2023.nc"  "2024.nc"  "2025.nc"
[127] "2026.nc"  "2027.nc"  "2028.nc"  "2029.nc"  "2030.nc"  "2031.nc"
[133] "2032.nc"  "2033.nc"  "2034.nc"  "2035.nc"  "2036.nc"  "2037.nc"
[139] "2038.nc"  "2039.nc"  "2040.nc"  "2041.nc"  "2042.nc"  "2043.nc"
[145] "2044.nc"  "2045.nc"  "2046.nc"  "2047.nc"  "2048.nc"  "2049.nc"
[151] "2050.nc"  "2051.nc"  "2052.nc"  "2053.nc"  "2054.nc"  "2055.nc"
[157] "2056.nc"  "2057.nc"  "2058.nc"  "2059.nc"  "2060.nc"  "2061.nc"
[163] "2062.nc"  "2063.nc"  "2064.nc"  "2065.nc"  "2066.nc"  "2067.nc"
[169] "2068.nc"  "2069.nc"  "2070.nc"  "2071.nc"  "2072.nc"  "2073.nc"
[175] "2074.nc"  "2075.nc"  "2076.nc"  "2077.nc"  "2078.nc"  "2079.nc"
[181] "2080.nc"  "2081.nc"  "2082.nc"  "2083.nc"  "2084.nc"  "2085.nc"
[187] "2086.nc"  "2087.nc"  "2088.nc"  "2089.nc"  "2090.nc"  "2091.nc"
[193] "2092.nc"  "2093.nc"  "2094.nc"  "2095.nc"  "2096.nc"  "2097.nc"
[199] "2098.nc"  "2099.nc"  "2100.nc"

ncfiles <- file.path(outdir, ncfiles_sub)

> ncfiles
  [1] "/data/sserbin/Modeling/dvmdostem/pecan_runs/salix_run.029/out/SA-median/-1901.nc"
  [2] "/data/sserbin/Modeling/dvmdostem/pecan_runs/salix_run.029/out/SA-median/1901.nc"
  [3] "/data/sserbin/Modeling/dvmdostem/pecan_runs/salix_run.029/out/SA-median/1902.nc"
  [4] "/data/sserbin/Modeling/dvmdostem/pecan_runs/salix_run.029/out/SA-median/1903.nc"
  [5] "/data/sserbin/Modeling/dvmdostem/pecan_runs/salix_run.029/out/SA-median/1904.nc"
  [6] "/data/sserbin/Modeling/dvmdostem/pecan_runs/salix_run.029/out/SA-median/1905.nc"
  [7] "/data/sserbin/Modeling/dvmdostem/pecan_runs/salix_run.029/out/SA-median/1906.nc"
  [8] "/data/sserbin/Modeling/dvmdostem/pecan_runs/salix_run.029/out/SA-median/1907.nc"
  [9] "/data/sserbin/Modeling/dvmdostem/pecan_runs/salix_run.029/out/SA-median/1908.nc"
 [10] "/data/sserbin/Modeling/dvmdostem/pecan_runs/salix_run.029/out/SA-median/1909.nc"
 [11] "/data/sserbin/Modeling/dvmdostem/pecan_runs/salix_run.029/out/SA-median/1910.nc"

nc_years <- as.numeric(gsub("^(-?[[:digit:]]{4})\\.nc", "\\1", ncfiles_sub))

> nc_years
  [1] -1901  1901  1902  1903  1904  1905  1906  1907  1908  1909  1910  1911
 [13]  1912  1913  1914  1915  1916  1917  1918  1919  1920  1921  1922  1923
 [25]  1924  1925  1926  1927  1928  1929  1930  1931  1932  1933  1934  1935
 [37]  1936  1937  1938  1939  1940  1941  1942  1943  1944  1945  1946  1947
 [49]  1948  1949  1950  1951  1952  1953  1954  1955  1956  1957  1958  1959
 [61]  1960  1961  1962  1963  1964  1965  1966  1967  1968  1969  1970  1971
 [73]  1972  1973  1974  1975  1976  1977  1978  1979  1980  1981  1982  1983
 [85]  1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  1994  1995
 [97]  1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  2007
[109]  2008  2009  2010  2011  2012  2013  2014  2015  2016  2017  2018  2019
[121]  2020  2021  2022  2023  2024  2025  2026  2027  2028  2029  2030  2031
[133]  2032  2033  2034  2035  2036  2037  2038  2039  2040  2041  2042  2043
[145]  2044  2045  2046  2047  2048  2049  2050  2051  2052  2053  2054  2055
[157]  2056  2057  2058  2059  2060  2061  2062  2063  2064  2065  2066  2067
[169]  2068  2069  2070  2071  2072  2073  2074  2075  2076  2077  2078  2079
[181]  2080  2081  2082  2083  2084  2085  2086  2087  2088  2089  2090  2091
[193]  2092  2093  2094  2095  2096  2097  2098  2099  2100
```